src/amp-viewer-host.js

```diff
@@ -22,5 +22,139 @@ import {
+   * looking at.
+   */
+  constructor(win, ampIframe, frameOrigin, messageHandler, opt_isWebview,
+    opt_isHandshakePoll) {
+    /** @const {!Window} */
+    this.win = win;
+    /** @private {!HTMLIFrameElement} */
+    this.ampIframe_ = ampIframe;
+    /** @private {function(string, *, boolean):(!Promise<*>|undefined)} */
+    this.messageHandler_ = messageHandler;
+    /** @const {boolean} */
+    this.isWebview_ = !!opt_isWebview;
+
+    if (this.isWebview_ || opt_isHandshakePoll) {
+      /** @private {number} */
+      this.pollingIntervalId_ = setInterval(this.initiateHandshake_.bind(
+        this, this.intervalCtr) , 1000); //poll every second
+    } else {
+      this.waitForHandshake_(frameOrigin);
+    }
+  }
+
+  /**
+   * @private
+   */
+  initiateHandshake_() {
+    this.log('initiateHandshake_');
+    if (this.ampIframe_) {
+      const channel = new MessageChannel();
+      let message = {
+        app: APP,
+        name: 'handshake-poll',
+      };
+      message = this.isWebview_ ? JSON.stringify(message) : message;
+      this.ampIframe_.contentWindow./*OK*/postMessage(
+        message, '*', [channel.port2]);
+
+      channel.port1.onmessage = function(e) {
+        const data = this.isWebview_ ? JSON.parse(e.data) : e.data;
+        if (this.isChannelOpen_(data)) {
+          this.win.clearInterval(this.pollingIntervalId_); //stop polling
+          this.log('messaging established!');
+          this.completeHandshake_(channel.port1, data.requestid);
+        } else {
+          this.messageHandler_(data.name, data.data, data.rsvp);
+        }
+      }.bind(this);
+    }
+  }
+
+  /**
+   * @param {string} targetOrigin
+   * @private
+   */
+  waitForHandshake_(targetOrigin) {
+    this.log('awaitHandshake_');
+    const listener = function(event) {
+      console.log('message!', event);
+      const target = this.ampIframe_.contentWindow;
+      if (event.origin == targetOrigin &&
+              this.isChannelOpen_(event.data) &&
+              (!event.source || event.source == target)) {
+        this.log(' messaging established with ', targetOrigin);
+        this.win.removeEventListener('message', listener);
+        const port = new WindowPortEmulator(this.win, targetOrigin, target);
+        this.completeHandshake_(port, event.data.requestid);
+      }
+    }.bind(this);
+    this.win.addEventListener('message', listener);
+  }
+
+  /**
+   * @param {!MessagePort|!WindowPortEmulator} port
+   * @param {string} requestId
+   * @private
+   */
+  completeHandshake_(port, requestId) {
+    let message = {
+      app: APP,
+      requestid: requestId,
+      type: MessageType.RESPONSE,
+    };
+
+    message = this.isWebview_ ? JSON.stringify(message) : message;
+    this.log('posting Message', message);
+    port./*OK*/postMessage(message);
+
+    this.messaging_ = new Messaging(this.win, port);
+    this.messaging_.setDefaultHandler(this.messageHandler_);
+
+    this.sendRequest('visibilitychange', {
```

Comment:
This is, looks like, an API of the `AmpViewer`, not something we should execute automatically ourselves.